### ISSUE TITLE
f

### ISSUE THREAD

**Issue by kimaram** [kimaram/chip#3]
ver2 백업

import shutil
import numpy as np
from PIL import Image
import torch
import time
import glob
import random
import os
import tkinter as tk
from tkinter import ttk, filedialog, messagebox
from collections import defaultdict
from torchvision import transforms
from efficientnet_pytorch import EfficientNet

# ---- 기존 모델 로직 ----
model_name = 'efficientnet-b0'
image_size = EfficientNet.get_image_size(model_name)
batch_size = 8
random_seed = 555
random.seed(random_seed)
torch.manual_seed(random_seed)

def run_sorting(app):
    since = time.time()
    defect_item = app.defect_var.get()
    if not defect_item:
        messagebox.showerror("오류", "불량 항목을 선택하세요.")
        return

    search_pattern = f"*_{defect_item}_Class_*.pt"
    matching_models = glob.glob(search_pattern)
    if not matching_models:
        messagebox.showerror("오류", f"{defect_item} 관련 모델 파일을 찾을 수 없습니다.\n(패턴: {search_pattern})")
        return

    model_path = os.path.abspath(matching_models[0])
    file_name = os.path.basename(model_path)

    try:
        class_cnt = int(file_name.split('_', 1)[0])
        defect_name = file_name.split('_', 2)[1]
        class_section = file_name.split("Class_")[1].replace(".pt", "")
        parts = class_section.split('_')
        class_names = {}
        for p in parts:
            num, name = p.split('.', 1)
            class_names[num] = name
        if len(class_names) != class_cnt:
            messagebox.showwarning("경고", f"클래스 수 불일치: {len(class_names)} vs {class_cnt}")
    except Exception as e:
        messagebox.showerror("오류", f"모델 파일명 파싱 실패:\n{file_name}\n\n{str(e)}")
        return

    if not os.path.exists(model_path):
        messagebox.showerror("오류", f"모델 파일이 존재하지 않습니다: {model_path}")
        return

    app.update_status(f"EfficientNet 모델 초기화 중... ({defect_name})")
    app.root.update()

    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    model = EfficientNet.from_name(model_name, num_classes=class_cnt).to(device)

    try:
        app.update_status("모델 가중치를 로드하는 중...")
        app.root.update()
        state_dict = torch.load(model_path, map_location=device)
        model.load_state_dict(state_dict)
    except Exception as e:
        messagebox.showerror("오류", f"모델 가중치 로드 실패: {str(e)}")
        return

    image_list = glob.glob(os.path.join(app.input_path, "**", "*.jpg"), recursive=True)
    if not image_list:
        messagebox.showinfo("정보", "분류할 이미지가 없습니다.")
        return

    save_path = app.output_path
    image_save_dirs = {name: os.path.join(save_path, name) for name in class_names.values()}
    for save_dir in image_save_dirs.values():
        os.makedirs(save_dir, exist_ok=True)

    tfms = transforms.Compose([
        transforms.Grayscale(num_output_channels=3),
        transforms.Resize(image_size),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ])

    total_images = len(image_list)
    results = defaultdict(int)
    model.eval()
    for i, image_file_path in enumerate(image_list):
        try:
            app.update_status(f"이미지 {i + 1}/{total_images} 분류 중...")
            app.progress_bar["value"] = (i + 1) / total_images * 100
            app.root.update_idletasks()

            img = tfms(Image.open(image_file_path)).unsqueeze(0).to(device)
            with torch.no_grad():
                outputs = model(img)
                idx = torch.topk(outputs, k=1).indices.squeeze(0).item()

            predicted_class_name = class_names.get(str(idx + 1), f"Unknown_{idx}")
            results[predicted_class_name] += 1
            shutil.copy(image_file_path, os.path.join(image_save_dirs[predicted_class_name], os.path.basename(image_file_path)))

        except Exception as e:
            app.update_status(f"Error: {str(e)}")
            continue

    time_elapsed = time.time() - since
    app.update_status(f'분류 완료! 소요 시간: {time_elapsed // 60:.0f}분 {time_elapsed % 60:.0f}초')

    analysis_text = f"--- Sorting 결과 분석 ({defect_name}) ---\n"
    for class_name, count in results.items():
        analysis_text += f"{class_name}: {count}개\n"

    app.display_results(analysis_text)
    messagebox.showinfo("완료", f"{defect_name} 이미지 분류가 완료되었습니다.")

class ImageSortingApp:
    def __init__(self, root):
        self.root = root
        self.root.title("🌿 D/L Image Auto Sorting Program Ver1.0")
        self.root.geometry("650x550")
        self.root.configure(bg="#E6F4EA")  # 연한 그린 배경

        self.input_path = ""
        self.output_path = ""
        self.defect_var = tk.StringVar()

        self.style = ttk.Style()
        self.apply_style()
        self.create_widgets()

    def apply_style(self):
        bg_color = "#E6F4EA"
        btn_color = "#9AD1B4"
        hover_color = "#7FC8A9"

        self.style.theme_use("clam")
        self.style.configure("TFrame", background=bg_color)
        self.style.configure("TLabel", background=bg_color, font=("맑은 고딕", 11))
        self.style.configure("TLabelframe", background=bg_color, font=("맑은 고딕", 11, "bold"))
        self.style.configure("TLabelframe.Label", background=bg_color, foreground="#3E6553")  # 제목 라벨 배경 통일
        self.style.configure("TButton",
                             font=("맑은 고딕", 11, "bold"),
                             foreground="#ffffff",
                             background=btn_color,
                             padding=8)
        self.style.map("TButton", background=[("active", hover_color)])
        self.style.configure("TCombobox", fieldbackground="#ffffff", background="#DFF7E1")
        self.style.configure("Horizontal.TProgressbar", troughcolor="#CFE9D6", background=hover_color)

    def create_widgets(self):
        main_frame = ttk.Frame(self.root, padding="20")
        main_frame.pack(fill=tk.BOTH, expand=True)

        title_label = ttk.Label(
            main_frame,
            text="🌸 D/L Image Auto Sorting Program 🌸",
            font=("Comic Sans MS", 16, "bold"),
            foreground="#4B7B57"
        )
        title_label.pack(pady=10)

        # 불량 항목 선택
        defect_frame = ttk.LabelFrame(main_frame, text="불량 항목 선택", padding="10", style="TLabelframe")
        defect_frame.pack(fill=tk.X, pady=10)
        combo = ttk.Combobox(defect_frame, textvariable=self.defect_var, state="readonly")
        combo['values'] = ('ChipOut', 'OLB', 'Film')
        combo.current(0)
        combo.pack(fill=tk.X, pady=3)

        # 입력 경로
        input_frame = ttk.LabelFrame(main_frame, text="분류할 이미지 경로", padding="10", style="TLabelframe")
        input_frame.pack(fill=tk.X, pady=10)
        self.input_entry = ttk.Entry(input_frame)
        self.input_entry.pack(side=tk.LEFT, fill=tk.X, expand=True, padx=(0, 5))
        ttk.Button(input_frame, text="📁 열기", command=self.select_input_path).pack(side=tk.RIGHT)

        # 출력 경로
        output_frame = ttk.LabelFrame(main_frame, text="분류 완료 이미지 경로", padding="10", style="TLabelframe")
        output_frame.pack(fill=tk.X, pady=10)
        self.output_entry = ttk.Entry(output_frame)
        self.output_entry.pack(side=tk.LEFT, fill=tk.X, expand=True, padx=(0, 5))
        ttk.Button(output_frame, text="📂 열기", command=self.select_output_path).pack(side=tk.RIGHT)

        # 시작 버튼
        start_btn = ttk.Button(main_frame, text="🌿 분류 시작 🌿", command=lambda: self.run_sorting_with_gui())
        start_btn.pack(pady=20, ipadx=10, ipady=5)

        # 상태 표시
        self.status_label = ttk.Label(main_frame, text="대기 중...", anchor="center", foreground="#2E8B57")
        self.status_label.pack(fill=tk.X, pady=(0, 5))

        # 프로그레스 바
        self.progress_bar = ttk.Progressbar(main_frame, orient="horizontal", mode="determinate", length=300)
        self.progress_bar.pack(fill=tk.X, pady=5)

        # 결과 분석
        analysis_frame = ttk.LabelFrame(main_frame, text="Sorting 결과 분석", padding="10", style="TLabelframe")
        analysis_frame.pack(fill=tk.BOTH, expand=True, pady=10)
        self.analysis_text = tk.Text(
            analysis_frame,
            height=10,
            state="disabled",
            wrap="word",
            bg="#F2FBF5",
            relief="flat",
            font=("맑은 고딕", 10)
        )
        self.analysis_text.pack(fill=tk.BOTH, expand=True)

    def select_input_path(self):
        path = filedialog.askdirectory()
        if path:
            self.input_path = path
            self.input_entry.delete(0, tk.END)
            self.input_entry.insert(0, path)

    def select_output_path(self):
        path = filedialog.askdirectory()
        if path:
            self.output_path = path
            self.output_entry.delete(0, tk.END)
            self.output_entry.insert(0, path)

    def update_status(self, message):
        self.status_label.config(text=message)
        self.root.update_idletasks()

    def display_results(self, text):
        self.analysis_text.config(state="normal")
        self.analysis_text.delete(1.0, tk.END)
        self.analysis_text.insert(tk.END, text)
        self.analysis_text.config(state="disabled")

    def run_sorting_with_gui(self):
        if not self.input_path or not self.output_path:
            messagebox.showerror("오류", "분류할 이미지 경로와 저장 경로를 모두 지정하세요.")
            return

        # 버튼 비활성화
        for child in self.root.winfo_children():
            if isinstance(child, ttk.Frame):
                for widget in child.winfo_children():
                    if isinstance(widget, ttk.Button):
                        widget.config(state="disabled")

        self.progress_bar["value"] = 0
        self.display_results("")

        # 백그라운드에서 분류 작업 시작
        self.root.after(100, lambda: run_sorting(self))

        # 작업 완료 후 버튼 활성화
        def re_enable_buttons():
            for child in self.root.winfo_children():
                if isinstance(child, ttk.Frame):
                    for widget in child.winfo_children():
                        if isinstance(widget, ttk.Button):
                            widget.config(state="normal")

        self.root.after(100, re_enable_buttons)

if __name__ == '__main__':
    root = tk.Tk()
    app = ImageSortingApp(root)
    root.mainloop()

# ---- 메인 실행 ----
if __name__ == '__main__':
    root = tk.Tk()
    app = ImageSortingApp(root)
    root.mainloop()
